src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingCrafter.java

```diff
@@ -173,4 +176,55 @@ public static boolean canBuildingCanLearnMoreRecipes(final int buildingLevel, fi
     {
         return (Math.pow(2, buildingLevel) * EXTRA_RECIPE_MULTIPLIER) >= (learnedRecipes + 1);
     }
+
+    /**
+     * @param token
+     * @return whether the recipe can bee added based on tokens.
+     */
+    protected Optional<Boolean> canRecipeBeAddedBasedOnTags(final IToken token)
```

Comment:
This seems like a duplicate. And why not just make it return true/false ?

Reply:
Because we need three cases:

- True means the recipe is allowed. 
- False means the recipe is explicitly denied. 
- Null means false unless a following rule allows it.

Reply:
But do we need it in the crafter + in the worker override?
(Also, we never return null in the method, that's why I asked why we don't return a boolean)

Reply:
Actually we probably don't need it in both places; I am testing that now. The actual `canRecipeBeAdded` method always returns true or false but the `canRecipeBeAddedBasedOnTags` needs to be able to say Yes, No, or "I don't care" before letting the builder-specific logic.

Reply:
I mean the canRecipeBeAddedBasedOnTags also atm always is true/false and never null

Reply:
or did you want to return nullable(null) instead of nullable(false) ?

Reply:
I return `Optional.ofNullable(false)` which is like Null but won't generate null pointer exceptions. You have to set it to something.

Reply:
```
    public static <T> Optional<T> ofNullable(T var0) {
        return var0 == null ? empty() : of(var0);
    }
```

Optional.ofNullable(false) is the same as Optional.of(false), you have to give it null